framework/src/constraints/AutomaticMortarGeneration.C

```diff
@@ -47,6 +49,159 @@
+
+      dof_map.dof_indices(elem, dof_indices_t1x, _t1x_var_num);
+      dof_map.dof_indices(elem, dof_indices_t1y, _t1y_var_num);
+      dof_map.dof_indices(elem, dof_indices_t1z, _t1z_var_num);
+
+      dof_map.dof_indices(elem, dof_indices_t2x, _t2x_var_num);
+      dof_map.dof_indices(elem, dof_indices_t2y, _t2y_var_num);
+      dof_map.dof_indices(elem, dof_indices_t2z, _t2z_var_num);
+
+      //
+
+      // For each node of the Elem, if it is in the secondary_node_to_nodal_normal
+      // container, set the corresponding nodal normal dof values.
+      for (MooseIndex(elem->n_vertices()) n = 0; n < elem->n_vertices(); ++n)
+      {
+        auto it = _amg._secondary_node_to_nodal_normal.find(elem->node_ptr(n));
+        if (it != _amg._secondary_node_to_nodal_normal.end())
+        {
+          _nodal_normals_system->solution->set(dof_indices_nnx[n], it->second(0));
+          _nodal_normals_system->solution->set(dof_indices_nny[n], it->second(1));
+          _nodal_normals_system->solution->set(dof_indices_nnz[n], it->second(2));
+        }
+
+        auto it_tangent = _amg._secondary_node_to_hh_nodal_tangents.find(elem->node_ptr(n));
+        if (it_tangent != _amg._secondary_node_to_hh_nodal_tangents.end())
+        {
+          _nodal_normals_system->solution->set(dof_indices_t1x[n], it_tangent->second[0](0));
+          _nodal_normals_system->solution->set(dof_indices_t1y[n], it_tangent->second[0](1));
+          _nodal_normals_system->solution->set(dof_indices_t1z[n], it_tangent->second[0](2));
+
+          _nodal_normals_system->solution->set(dof_indices_t2x[n], it_tangent->second[1](0));
+          _nodal_normals_system->solution->set(dof_indices_t2y[n], it_tangent->second[1](1));
+          _nodal_normals_system->solution->set(dof_indices_t2z[n], it_tangent->second[1](2));
+        }
+
+      } // end loop over nodes
+    }   // end loop over elems
+
+    // Finish assembly.
+    _nodal_normals_system->solution->close();
+
+    std::set<std::string> sys_names = {nodal_normals_sys_name};
+
+    // Write the nodal normals to file
+    ExodusII_IO nodal_normals_writer(_amg._mesh);
+
+    // Default to non-HDF5 output for wider compatibility
+    nodal_normals_writer.set_hdf5_writing(false);
+
+    nodal_normals_writer.write_equation_systems(
+        "nodal_geometry_only.e", nodal_normals_es, &sys_names);
+  }
+
+private:
+  /// The mortar generation object that we will query for nodal normal and tangent information
+  AutomaticMortarGeneration & _amg;
+
+  ///@{
+  /** Member variables for geometry debug output */
+  libMesh::System * _nodal_normals_system = nullptr;
+  unsigned int _nnx_var_num;
+  unsigned int _nny_var_num;
+  unsigned int _nnz_var_num;
+
+  unsigned int _t1x_var_num;
+  unsigned int _t1y_var_num;
+  unsigned int _t1z_var_num;
+
+  unsigned int _t2x_var_num;
+  unsigned int _t2y_var_num;
+  unsigned int _t2z_var_num;
+  ///@}
+};
+
+registerMooseObject("MooseApp", MortarNodalGeometryOutput);
```

Comment:
I can't see a test file requesting this output. Is it supposed to?

Reply:
Any input with `debug_mesh = true` in a mortar constraint will implicitly create this output object. You have one input in the contact module that does this ... otherwise I wouldn't have had to create this hahaha. But yea if we don't see coverage for this in the framework testing, I will add a framework level test requesting debug output

Reply:
My understanding: You are creating an output object that's registered. But I don't think a user can leverage it from the input file, despite being registered. It's still driven from the `debug_mesh` flag. May be potentially confusing to the user?

Reply:
Yea I can see that. I will think about how to do this best without doing any registration